Make the new options module fully type safe

dcbaker opened the pull request:
This fixes all of the issues in the options module, as well as fixing type safety issues introduced by it. This really falls into four parts:

1. fix the pre-existing issues with the `UserOption` classes
2. clean up some uses of `Any` that would hide real issues
3. Correct annotations for the OptionKey type that were broken
4. add missing annotations to the `OptionStore` class.

This fully supersedes #13616

dnicolodi commented on mesonbuild/options.py:
```diff
@@ -141,7 +141,7 @@ def __setstate__(self, state: T.Dict[str, T.Any]) -> None:
         return self._hash
 
-    def _to_tuple(self) -> T.Tuple[str, str, str, MachineChoice, str]:
+    def _to_tuple(self) -> T.Tuple[str, MachineChoice, str]:
```
The commit message mentions `_as_tuple` but the commit changes the annotation of `_to_tuple()`

dnicolodi commented on mesonbuild/options.py:
```diff
@@ -324,7 +332,10 @@ def __post_init__(self, value_: int) -> None:
+        self.__choices: str = ', '.join(choices)
+
+    def printable_choices(self) -> T.Optional[T.List[str]]:
+        return [self.__choices]
```
Overloading `printable_choices()` to return a description of the valid range is a bit weird. The use of `self.__choices` makes it even less understandable. Would it make sense to move the computation of the string returned by `printable_choices()` into the method itself? I don't think that this code path is performance sensitive and requires caching the computation, but if it does, decorating the method with `@functools.cache` may be a better solution.

dcbaker replied:
This is super important actually, because choices needs to be a `Container[T]`, matching the type of the `UserOption[T]`, but printable choices needs to be `List[str] | None`, regardless of what `T` is. The entire point of `printable_choices` is to fix the fact that we both need a string definable set of choices (like those used in Combo and Array options that are essentially a set of valid values, and the choices of integer options, where the constraints are a range of valid values. The status quo is that `IntegerOption` shoves `'>=0, < 100'` into `choices` and acts like it's cool, but then `BooleanOption` puts `[True, False]` in there.

dnicolodi commented on mesonbuild/coredata.py:
```diff
@@ -599,7 +599,7 @@ def update_project_options(self, project_options: 'MutableKeyedOptionDictType',
             if type(oldval) is not type(value):
                 self.optstore.set_value(key, value.value)
-            elif oldval.choices != value.choices:
+            elif oldval.printable_choices() != value.printable_choices():
```
Using `printable_choices()` for this is a bit strange. I think it is done to accommodate for `UserIntegerOption` which AFAICT is the only option type that specializes `printable_choices()` to return something not computed from its `choices` member.

dnicolodi replied:
I haven't checked if something else would break, but having `UserIntegerOption` return a `slice` object for its `choice` property would solve this maybe a bit more elegantly. I first thought about using a `range` object, but `range` needs both a lover and an upper limit. `slice` does not. The alternative would be to return a custom object:
```python
@dataclasses.dataclass
class Range[_T]:

    min_value: _T
    max_value: _T

    def __contains__(self, x):
        return self.min_value <= x <= self.max_value

    def __str__(self):
        s = []
        if self.min_value is not None:
            s.append(f'>= {self.min_value!s}')
        if self.max_value is not None:
            s.append(f'<= {self.max_value!s}')
        return ', '.join(s)
```
The implementation of `UserIntegerOption` could be reworked to store a `Range[int]` instead than `min_value` and `max_value` directly.

dnicolodi commented on mesonbuild/options.py:
```diff
@@ -300,78 +319,85 @@ def validate_value(self, value: T.Any) -> bool:
+class UserIntegerOption(_UserIntegerBase[int]):
+
+    min_value: T.Optional[int] = None
+    max_value: T.Optional[int] = None
```
Given that this class derives from `_UserIntegerBase`, do the dataclass members need to be declared again?

dcbaker replied:
I made a choice to not implement them in the `_UserIntegerBase` because the two classes have different default values, which even still, the shared base is a bit of a compromise with code reuse, since technically the Umask option always has a min and max.

dnicolodi replied:
I see what you mean. Rading the code, when I found `min_value: T.Optional[_T]` in the base class, my brain autocompleted it appending a `= None` as it usually follows members declared `T.Optional`. Thus the redefinition here seemed redundant.

dcbaker commented:
@dnicolodi: I've added an explicit method for checking the choices, so all of that is hidden inside the options module, which I think is a bit more elegant than having coredata poke at the internals of the `UserOption` classes. I've also added a `Range[T]` implementation. I'm not sure on it, it does make some things nicer, but its more code and makes other things more complex, so for now I haven't folded into the series itself, but left it separate so it can be evaluated separately.

dnicolodi commented:
The suggestion to add the `Range` type was done in the optic of allowing to compare the `choice` member of enumerated options types and int option types transparently, thus it is somehow alternative to your `choices_are_different()` function. It does not add much value otherwise.

dnicolodi commented on mesonbuild/options.py:
```diff
@@ -483,7 +483,31 @@ def is_auto(self) -> bool:
+    """
+    if isinstance(a, EnumeratedUserOption):
+        # If this isn't hte case the above failed somehow
+        assert isinstance(b, EnumeratedUserOption), 'for mypy'
```
Is mypy really this dumb? Isn't `isinstance()` recognized as a type narrowing check? I don't know what else is then...

dcbaker replied:
It is a type narrowing check, but only for `a`, `b` is still left as `_U`, and even when I tried adding things like:
```python
assert type(a) is type(b)
```
mypy wasn't smart enough to figure out that if `a` is `EnumeratedUserOption` that `b` would also be `EnumeratedUserOption`.

dnicolodi replied:
Oh, sorry, I misunderstood how type vars work when they reference unions. I though that if the same type vars appears in a function signature it is forced to resolve to the same concrete type. Evidently it does not work like this.

dcbaker replied:
Maybe it should? It would in rust and C++ wouldn't it? Maybe I should open a typing issue about this.

dnicolodi commented on mesonbuild/options.py:
```diff
@@ -483,7 +483,31 @@ def is_auto(self) -> bool:
+    :return: True if the choices have changed, otherwise False
+    """
+    if isinstance(a, EnumeratedUserOption):
+        # If this isn't hte case the above failed somehow
```
Typo: `s/hte/the/`

dcbaker commented:
@dnicolodi fixed the typo, and dropped the range patch. I played around with getting better error messages, and just felt like it wasn't worth the amount of code involved.

bruchar1 commented on mesonbuild/compilers/cpp.py:
```diff
@@ -741,24 +749,28 @@ def get_option_link_args(self, options: 'KeyedOptionDictType') -> T.List[str]:
+        key = self.form_compileropt_key('winlibs')
+        opts[key] = options.UserStringArrayOption(
+            self.make_option_name(key),
+            'Windows libs to link against.',
```
This should be `Standard Win libraries to link against` like other places

dcbaker replied:
That is copy and pasted from one place to another, but I can standardize that as a follow on patch

dcbaker replied:
Oh, never mind, it was changed at one point.

bruchar1 reviewed (commented):
This isn't directly related to this PR, but there is a lot of repetition for defining options, especially for the `winlibs` option. I wonder if a global dict or function should be used to prevent having a different description depending where the option is defined?

bruchar1 commented on mesonbuild/compilers/cpp.py:
```diff
@@ -393,15 +395,15 @@ def __init__(self, ccache: T.List[str], exelist: T.List[str], version: str, for_
+        key = self.form_compileropt_key('eh')
+        opts[key] = options.UserComboOption(
+            self.make_option_name(key),
+            'C++ exception handling type.',
```
nitpick: this one ends with a `.`, while other options don't...

dcbaker replied:
That has apparently always existed. The main branch right now has that peroid

bruchar1 commented on mesonbuild/options.py:
```diff
@@ -529,19 +516,21 @@ class BuiltinOption(T.Generic[_T]):
         if value is None:
             value = self.prefixed_default(name, prefix)
-        keywords = {'yielding': self.yielding, 'value': value}
+        keywords = {'yielding': self.yielding, 'value_': value}
```
Where does that `value_` comes from?

dcbaker replied:
Because to use dataclasses I had to rename the argument from `value` to `value_`, otherwise you can't convert the type. See https://github.com/python/mypy/issues/15423, where I argued this was a bug but upstream argued it was the intended behavior.

eli-schwartz replied:
Ah, but python doesn't *need* convertors for dataclasses, do we...

dcbaker replied:
I'd laugh if the whole thing didn't make me so mad

bruchar1 commented on mesonbuild/mintro.py:
```diff
@@ -316,15 +316,15 @@ def add_keys(opts: 'T.Union[dict[OptionKey, UserOption[Any]], cdata.KeyedOptionD
                     optdict['choices'] = c
             else:
-                raise RuntimeError("Unknown option type")
+                raise RuntimeError("Unknown option type: ", type(opt))
```
Use single quotes

bruchar1 reviewed (commented):
Overall, looks good!

eli-schwartz commented:
I cherry-picked a few (more) of these to master. Haven't taken a close look at the rest and it's getting late on Friday.

eli-schwartz commented on mesonbuild/compilers/compilers.py:
```diff
@@ -35,9 +35,10 @@
     UserOptionType = T.TypeVar('UserOptionType', bound=options.UserOption)
 
+_T = T.TypeVar('_T')
+
```
This diff hunk confused me because it seems to be in the wrong commit.

dcbaker replied:
Moved to the correct commit

eli-schwartz commented on mesonbuild/compilers/mixins/emscripten.py:
```diff
@@ -63,7 +63,8 @@ def get_options(self) -> coredata.MutableKeyedOptionDictType:
             'Number of threads to use in web assembly, set to 0 to disable',
-            (0, None, 4))  # Default was picked at random
+            4,  # Default was picked at random
+            min_value=0)
```
I think this may be a good case where it's worth explaining the change in the commit message. After thinking about this, I think that this works because the dataclass also fits a historic mess where the "value" member isn't consistently used by all option types, when we do stuff like overloading it with minimum/maximum values instead?

jpakkane commented:
This probably conflicts with option refactor quite badly. Which one should be merged first?

dcbaker commented:
I think this one, and then following that: https://github.com/mesonbuild/meson/pull/13737. Which would give us a really solid base of typing guarantees to build the option refactor on top of so that we have a fairly good idea that things work across the changes. I've actually started doing that rebase on top of 13737, and I've found several issues already. I'll try to push that work upstream somewhere we can look at it after a bit, right now I need to go start dinner.

bruchar1 commented on mesonbuild/options.py:
```diff
@@ -265,6 +265,11 @@ def printable_value(self) -> T.Union[str, int, bool, T.List[T.Union[str, int, bo
+    def printable_choices(self) -> T.Optional[T.List[str]]:
+        if not self.choices:
+            return None
+        return [str(c) for c in self.choices]
```
It seems that `self.choices` could be a `str`. In that case, that would result in a list of letters, isn't it?

dcbaker replied:
Yeah, at the point it could. So I'll fix that.

bruchar1 commented on mesonbuild/compilers/rust.py:
```diff
@@ -220,11 +220,16 @@ def use_linker_args(cls, linker: str, version: str) -> T.List[str]:
+        opts[key] = options.UserComboOption(
+            self.make_option_name(key),
+            'Rust edition to use',
+            ['none', '2015', '2018', '2021'],
```
2024 is dropped?

dcbaker replied:
Rebase error. 2024 wasn't in Meson when this was originally written :D

bruchar1 commented on mesonbuild/options.py:
```diff
@@ -445,13 +431,12 @@ def extend_value(self, value: T.Union[str, T.List[str]]) -> None:
-        self.name: T.Optional[str] = None  # TODO: Refactor options to all store their name
+    choices: T.List[str] = dataclasses.field(
+        # Ensure we get a copy with the lambda
+        default_factory=lambda: ['enabled', 'disabled', 'auto'], init=False)
```
Why not simply generate is in `__post_init__` instead?

dcbaker replied:
I dislike using `__post_init__`, tbh. It feels like a code smell to me. IMO, it's also distraction because it moves from "here are all of the fields for my class in one place" to "fields for my class are declared in multiple places

bruchar1 replied:
one way or other are ok for me.

bruchar1 commented on mesonbuild/optinterpreter.py:
```diff
@@ -272,10 +272,11 @@ def string_array_parser(self, name: str, description: str, args: T.Tuple[bool, _
                 raise mesonlib.MesonException('Value does not define an array: ' + value)
         # XXX: the value of choices is correct, the annotation is wrong.
         #      the annotation will be fixed in a later commit
```
Is this comment still relevant?

dcbaker replied:
No it should be dropped.

bruchar1 commented on mesonbuild/mintro.py:
```diff
@@ -317,15 +317,15 @@ def add_keys(opts: 'T.Union[dict[OptionKey, UserOption[Any]], cdata.KeyedOptionD
                     optdict['choices'] = c
             else:
-                raise RuntimeError("Unknown option type")
+                raise RuntimeError('Unknown option type: ', type(opt))
```
Is it supported to pass a type as second argument of RuntimeError?

dcbaker replied:
Of course, something like a `match` statement that was required to be exhaustive would be even better

jpakkane commented:
How ready is this for merging?

dcbaker commented:
From my perspective I've addressed all of the review feedback, with the exception of the question about `__post_init__` vs `init=False`. I have a preference but if the wider consensus is to go the other way I can do that.

jpakkane commented:
I don't really have a preference on that, I just want this in so optionrefactor can proceed.

jpakkane commented:
Lint failure is unrelated.
